src/coreclr/jit/valuenum.cpp

```diff
@@ -8807,14 +8801,18 @@ void Compiler::fgValueNumberTree(GenTree* tree)
             if (tree->OperIs(GT_ARR_LENGTH))
             {
                 // Case 1: ARR_LENGTH(FROZEN_OBJ)
-                ValueNum addressVN = tree->gtGetOp1()->gtVNPair.GetLiberal();
+                ValueNum addressVN = vnStore->VNNormalValue(tree->gtGetOp1()->gtVNPair.GetLiberal());
                 if (vnStore->IsVNHandle(addressVN) && (vnStore->GetHandleFlags(addressVN) == GTF_ICON_OBJ_HDL))
                 {
                     size_t handle = vnStore->CoercedConstantValue<size_t>(addressVN);
                     int    len    = this->info.compCompHnd->getArrayOrStringLength((CORINFO_OBJECT_HANDLE)handle);
                     if (len >= 0)
                     {
+                        ValueNumPair op1vnp;
+                        ValueNumPair op1Xvnp;
+                        vnStore->VNPUnpackExc(tree->gtGetOp1()->gtVNPair, &op1vnp, &op1Xvnp);
                         tree->gtVNPair.SetBoth(vnStore->VNForIntCon(len));
+                        tree->gtVNPair = vnStore->VNPExcSetUnion(tree->gtVNPair, op1Xvnp);
                         return;
```

Comment:
The early return here will make the tree not appear in the dump. Early returns are not meant to be used `fgValueNumberTree`. I would once again advise to move this folding logic into a callee of `VNForFunc(unary)` (it should be about the same TP-wise since we are only looking for VNs which will always have the same liberal and conservative components here).

Reply:
@SingleAccretion thanks, move to unary, anything else?